docs/configuration.rst
.. _configuration:

Configuration
=============

The configuration file is in the JSON format. It consists of nested
key-value pairs.

For example::

    {
        "json_state_file_path": "/var/lib/pghoard/pghoard_state.json"
        "backup_sites": {
            "mycluster": {
                "nodes": [
                    {
                        "host": "127.0.0.1",
                        "password": "secret",
                        "port": 5432,
                        "user": "backup",
                        "slot": "pghoard"
                    }
                ],
                "basebackup_count": 5,
                "basebackup_mode": "delta",
                "object_storage": {
                    "storage_type": "local",
                    "directory": "/tmp/pghoard/backups"
                }
            }
        }
    }

Global Configuration
--------------------

Global configuration options are specified at the top-level.
In this documentation we group them by categories.


Generic Configuration
~~~~~~~~~~~~~~~~~~~~~




active (default ``true``)
        Can also be set on the ``backup_site`` level to disable taking of new backups
        and to stop the deletion of old ones
backup_location
        Where ``pghoard`` will create its internal data structures for local
        state data.
hash_algorithm (default ``"sha1"``)
        The hash algorithm used for calculating checksums for WAL or other files. Must
        be one of the algorithms supported by Python's `hashlib <https://docs.python.org/3/library/hashlib.html#hash-algorithms>`_
json_state_file_path (default ``"/var/lib/pghoard/pghoard_state.json"``)
        Location of the JSON state file path which describes the state of the
        ``pghoard`` process.
maintenance_mode_file (default ``"/var/lib/pghoard/maintenance_mode_file"``)
                        Trigger file for maintenance mode: if a file exists at
                        this location no new backup actions will be started)
                        FIXME: define "new backup actions"
transfer (default see below)
  A JSON object defining the WAL/basebackup transfer parameters.

  Example::

    {
      transfer: {
        thread_count: 4,
        upload_retries_warning_limit: 3
      }
    }

  thread_count (default ``min(cpu_count + 3, 20)``)
    Number of parallel uploads / downloads
  upload_retries_warning_limit (default ``3``)
    Create an alert file ``upload_retries_warning`` after this many failed
    upload attempts. See (FIXME: link to alert system)
tar_executable (default ``"pghoard_gnutaremu"``)
  The tar command to use for restoring basebackups. This must be GNU tar because some
  advanced switches like ``--transform`` are needed. If this value is not defined (or
  is explicitly set to ``"pghoard_gnutaremu"``), Python's internal tarfile
  implementation is used. The Python implementation is somewhat slower than the
  actual tar command and in environments with fast disk IO (compared to available CPU
  capacity) it is recommended to set this to ``"tar"``.
restore_prefetch (default ``transfer.thread_count``)
  Number of files to prefetch when performing archive recovery.  The default
  is the number of Transfer Agent threads to try to utilize them all.


.. _configuration_logging:

Logging configuration
~~~~~~~~~~~~~~~~~~~~~

log_level (default ``"INFO"``)
            Determines log level of ``pghoard``.
syslog (default ``false``)
  Enable / disable syslog logging
syslog_address (default ``"/dev/log"``)
  Determines syslog address to use in logging (requires syslog to be true as
  well)
syslog_facility (default ``"local2"``)
  Determines syslog log facility. (requires syslog to be true as well)


.. _configuration_monitoring:

Monitoring
~~~~~~~~~~

alert_file_dir (default ``backup_location`` if set else ``os.getcwd()``)
  Directory in which alert files for replication warning and failover are
  created.
stats (default ``null``)
  When set, enables sending to a statsd daemon that supports Telegrag or DataDog
  syntax with tags.
  The value is a JSON object, for example::

    {
        "host": "<statsd address>",
        "port": <statsd port>,
        "format": "<statsd message format>",
        "tags": {
            "<tag>": "<value>"
        }
    }

  host
    The statsd host address
  port
    The statsd listening port
  format (default ``"telegraf"``)
      Determines statsd message format. Following formats are supported:

        - ``telegraf`` `Telegraf spec <https://github.com/influxdata/telegraf/tree/master/plugins/inputs/statsd>`_
        - ``datadog`` `DataDog spec <http://docs.datadoghq.com/guides/dogstatsd/#datagram-format>`_

  :tags: (default null)
      The tag key can be used to enter optional tag values for the metrics
push_gateway (default ``null``)
  When set, enables sending metrics to a Prometheus Pushgateway with tags.
  The value is a JSON object, for example::

    {
        "endpoint": "<pushgateway address>",
        "tags": {
            "<tag>": "<value>"
        }
    }

  endpoint
    The pushgateway address
  tags
    An object mapping tags to their values.


.. _configuration_http:

HTTP Server configuration
~~~~~~~~~~~~~~~~~~~~~~~~~

The pghoard daemon needs to listen on an HTTP port for the archive command and
for fetching of basebackups/WAL's when restoring if not using an object store.

http_address (default ``"127.0.0.1"``)
               Address to bind the PGHoard HTTP server to.  Set to an empty string to
               listen to all available IPv4 addresses.   Set it to the IPv6 ``::`` wildcard
               address to bind to all available IPv4 and IPv6 addresses.
http_port (default ``16000``)
            HTTP webserver port. Used for the archive command and for fetching of
            basebackups/WAL's when restoring if not using an object store.




.. _configuration_compression:

Compression
~~~~~~~~~~~

The PostgreSQL write-ahead log (WAL) and basebackups are compressed with
Snappy (default), Zstandard (configurable, level 3 by default) or LZMA (configurable,
level 0 by default) in order to ensure good compression speed and relatively small backup size.
For performance critical applications it is recommended to test compression
algorithms to find the most suitable trade-off for the particular use-case.
E.g. Snappy is fast but yields larger compressed files, Zstandard (zstd) on the other hand
offers a very wide range of compression/speed trade-off.

The top-level ``compression`` key allows to define compression options::

  {
    "compression": {
      "algorithm": "snappy",
      "level": 3,
      "thread_count": 4
    }
  }

algorithm (default ``snappy``)
            The compression algorithm to use. Available algorithms are
            ``snappy``, ``zstd``, and ``lzma``
level (default ``0`` for ``lzma`` and ``zstd``, ``3`` for ``snappy``)
        The compression level to use. Depends on the algorithm used.
thread_count (default to ``cpu_count`` + 1)
        The number of threads used for parallel compression.
        Contrary to ``basebackup_compression_threads`` this is the number of
        compression threads started by ``pghoard``, not internal compression
        threads for libraries supporting it, and is then applicable to any
        compression algorithm.


Backup sites
------------

The key ``backup_sites`` contains configuration for groups of PostgreSQL clusters (here
called ``sites``). Each backup site configures how to backup the different nodes
it comprises. Each site can be configured separately, under an idenfiying
site name (example: ``mysite``).

A backup site contains an array of at least one node. For each node, the connection
information is required. The keys for a node are libpq parameters, for example::

  {
    "backup_sites": {
      "mysite": {
        "nodes": [
            {
                "host": "127.0.0.1",
                "password": "secret",
                "port": 5432,
                "user": "backup",
                "slot": "pghoard",
                "sslmode": "require"
            }
        ]
      }
    }
  }

It is advised to use a replication slot when performing a form of wal streaming archiving (``pg_receivexlog`` or ``walreceiver`` modes).

nodes (no default)
  A node can be described as an object of libpq key: value connection info pairs or libpq
  connection string or a ``postgres://`` connection uri. If for example you'd
  like to use a streaming replication slot use the syntax {... "slot": "slotname"}.
pg_data_directory (no default)
  This is used when the ``local-tar`` or ``delta`` ``basebackup_mode`` is in
  use. The data directory must point to PostgreSQL's ``$PGDATA`` and must be readable by the
  ``pghoard`` daemon.
prefix: (default site_name)
  Path prefix to use for all backups related to this site.
pg_bin_directory: (default find binaries from well-known directories)
  Where to find the ``pg_basebackup`` and ``pg_receivewal`` (``pg_receivexlog``
  for PG < 10).
  If a value is not supplied, ``pghoard`` will attempt to find matching binaries
  from various well-known locations. If ``pg_data_directory`` is set and points to a
  valid data directory the lookup is restricted to the version contained in
  the given data directory.


.. _configuration_basebackup:

Basebackup configuration
~~~~~~~~~~~~~~~~~~~~~~~~

The following options all concern various aspect of the basebackup process and
their retention policy.

basebackup_mode (default ``"basic"``)
  The way basebackups should be created. We support 4 different modes, the first
  two use ``pg_basebackup`` while the rest directly read the files from the
  cluster. Neither ``basic`` nor ``pipe`` modes support multiple tablespaces.

  ``basic``
    runs ``pg_basebackup`` and waits for it to write an uncompressed tar file on the
    disk before compressing and optionally encrypting it.
  ``pipe``
    pipes the data directly from ``pg_basebackup`` to PGHoard's
    compression and encryption processing reducing the amount of temporary disk
    space that's required.
  ``local-tar``
    Can be used only when running on the same host as the
    PostgreSQL cluster. Instead of using ``pg_basebackup``, PGHoard reads the files directly from ``$PGDATA`` in this mode and compresses and optionally encrypts them.  This mode allows backing up user
    tablespaces. Note that the ``local-tar`` backup mode can not be used on replica servers
    prior to PostgreSQL 9.6 unless the pgespresso extension is installed.

  ``delta``
    similar to ``local-tar``, but only changed files are uploaded into the storage.
    On every backup snapshot of the data files is taken, this results in a manifest file,
    describing the hashes of all the files needed to be backed up.
    New hashes are uploaded to the storage and used together with complementary
    manifest from control file for restoration.

  In order to properly assess the efficiency of ``delta`` mode in comparison with
  ``local-tar``, one can use ``local-tar-delta-stats`` mode, which behaves the same as
  ``local-tar``, but also collects the metrics as if it was ``delta`` mode. It can help
  in decision making of switching to ``delta`` mode.
basebackup_thread (default ``1``)
  How many threads to use for tar, compress and encrypt tasks. Only applies for
  ``local-tar`` basebackup mode. Only values 1 and 2 are likely to be sensible for
  this, with higher thread count speed improvement is negligible and CPU time is
  lost switching between threads.






The following options define how to schedule basebackups.

basebackup_interval_hours (default ``24``)
  How often to take a new basebackup of a cluster.  The shorter the interval,
  the faster your recovery will be, but the more CPU/IO usage is required from
  the servers it takes the basebackup from.  If set to a null value basebackups
  are not automatically taken at all.
basebackup_hour (default undefined)
  The hour of day during which to start new basebackup. If backup interval is
  less than 24 hours this is the base hour used to calculate the hours at which
  backup should be taken. E.g. if backup interval is 6 hours and this value is
  set to 1 backups will be taken at hours 1, 7, 13 and 19. This value is only
  effective if also ``basebackup_interval_hours`` and ``basebackup_minute`` are
  set.
basebackup_minute (default undefined)
  The minute of hour during which to start new basebackup. This value is only
  effective if also ``basebackup_interval_hours`` and ``basebackup_hour`` are
  set.


basebackup_chunks_in_progress (default  ``5``)
  How many basebackup chunks can there be simultaneously on disk while
  it is being taken. For chunk size configuration see ``basebackup_chunk_size``.
basebackup_chunk_size (default ``2147483648``)
  In how large backup chunks to take a ``local-tar`` basebackup. Disk space
  needed for a successful backup is ``basebackup_chunk_size *
  basebackup_chunks_in_progress``.
basebackup_compression_threads (default ``0``)
  Number of threads to use within compression library during basebackup. Only
  applicable when using compression library that supports internal multithreading,
  namely zstd at the moment. Default value ``0`` means not to use multithreading.

The following options manage the retention policy.

basebackup_age_days_max (default ``null``)
  Maximum age for basebackups. Basebackups older than this will be removed. By
  default this value is not defined and basebackups are deleted based on total
  count instead.
basebackup_count (default ``2``)
  How many basebackups should be kept around for restoration purposes.  The
  more there are the more diskspace will be used. If ``basebackup_max_age`` is
  defined this controls the maximum number of basebackups to keep; if backup
  interval is less than 24 hour or extra backups are created there can be more
  than one basebackup per day and it is often desirable to set
  ``basebackup_count`` to something slightly higher than the max age in days.
basebackup_count_min (default ``2``)
  Minimum number of basebackups to keep. This is only effective when
  ``basebackup_age_days_max`` has been defined. If for example the server is
  powered off and then back on a month later, all existing backups would be very
  old. However, in that case it is usually not desirable to immediately delete
  all old backups. This setting allows specifying a minimum number of backups
  that should always be preserved regardless of their age.



.. _configuration_archiving:

Archiving configuration
~~~~~~~~~~~~~~~~~~~~~~~


active_backup_mode (default ``pg_receivexlog``)
  Can be either ``pg_receivexlog`` or ``archive_command``. If set to
  ``pg_receivexlog``, ``pghoard`` will start up a ``pg_receivexlog`` process to be
  run against the database server.  If ``archive_command`` is set, we rely on the
  user setting the correct ``archive_command`` in
  ``postgresql.conf``. You can also set this to the experimental ``walreceiver`` mode
  whereby pghoard will start communicating directly with PostgreSQL
  through the replication protocol. (Note requires psycopg2 >= 2.7)


pg_receivexlog
  When active backup mode is set to ``"pg_receivexlog"`` this object may
  optionally specify additional configuration options. The currently available
  options are all related to monitoring disk space availability and optionally
  pausing xlog/WAL receiving when disk space goes below configured threshold.
  This is useful when PGHoard is configured to create its temporary files on
  a different volume than where the main PostgreSQL data directory resides. By
  default this logic is disabled and the minimum free bytes must be configured
  to enable it.

  Example::

    {
      "backup_sites": {
        "mysite": {
          "pg_receivexlog": {
            "disk_space_check_interval": 10,
            "min_disk_free_bytes": null,
            "resume_multiplier": 1.5
          }
        }
      }

  :disk_space_check_interval: (default ``10``)
      How often (in seconds) to check available disk space.
  :min_disk_free_bytes: (default ``null``)
      Minimum bytes (in integer) that must be available in order to keep receiving
      xlogs/WAL from PostgreSQL. If available disk space goes below this
      limit a ``STOP`` signal is sent to the ``pg_receivexlog`` / ``pg_receivewal``
      application.
  :resume_multiplier: (default ``1.5``)
      Number of times the ``min_disk_free_bytes`` bytes of disk space that is
      required to start receiving xlog/WAL again (i.e. send the ``CONT`` signal to
      the ``pg_receivexlog`` / ``pg_receivewal`` process). Multiplier above 1
      should be used to avoid stopping and continuing the process constantly.



.. _configuration_restore:

Restore configuration
---------------------






.. _configuration_storage:

Storage configuration
~~~~~~~~~~~~~~~~~~~~~

FIXME: reformat that according to what's been done above

``object_storage`` (no default)

Configured in ``backup_sites`` under a specific site.  If set, it must be an
object describing a remote object storage.  The object must contain a key
``storage_type`` describing the type of the store, other keys and values are
specific to the storage type.

``proxy_info`` (no default)

Dictionary specifying proxy information. The dictionary must contain keys ``type``,

NEWS
pghoard DEV
===========

* Support for PostgreSQL version 14
* Major improvements for delta backup support
* Improve robustness of the Swift backend
* Testing and CI improvements

pghoard 2.2.0 (2021-04-12)
==========================

* Support for PostgreSQL versions 12 and 13
* Support for more flexible basebackup scheduling
* Testing and CI improvements
* Initial support for delta basebackups
* Security fixes for tar extraction

pghoard 2.1.0 (2019-05-27)
==========================

* Support for passing statistics to Prometheus
* Progress monitoring of basebackups
* Fix incompatibility with legacy PGHoard backups (2.0.0 was unable to
  restore backups taken with much older versions)
* Compatibility improvements for different S3 api implementations
* Support for Zstandard (zstd) compression
* Support SFTP file storage
* File storage APIs (rohmu) support streaming

pghoard 2.0.0 (2018-11-08)
==========================

* Support for PostgreSQL 11
* Drop support for PostgreSQL 9.2
* Multiprocess parallel basebackup restore
* Performance improvements
* Bug fixes
* Python 3.7 compatibility

pghoard 1.7.0 (2018-02-09)
==========================

* Support for PostgreSQL 10.1+
* Use fast checkpoints by default to make the start of backups faster.  Slow
  progress at the beginning of the backup has a tendency to confuse people.
* Allow overriding the name of the folder for a site's backups by using a
  new site-level configuration option ``prefix``.  Previously backups were
  always stored in a folder named after the site, optionally prefixed by an
  undocumented top-level configuration option ``path_prefix`` which is now
  deprecated.
* ``archive_cleanup`` storage errors are no longer fatal (just logged as
  errors now).  Total WAL segment storage size savings is reported after
  cleanup.  There is a 'dry run' mode available now to estimate cleanup
  effect before deleting actual items.
* Use botocore rather than boto for interfacing with S3. Botocore is the
  underlying and maintained library under the newer boto3 SDK.
* Add a new --tablespace-base-dir option to restore so you can put all
  the tablespaces under a single directory hierarchy.
* Miscellaneous bug fixes and improvements

pghoard 1.6.0 (2017-10-12)
==========================

* Support for PostgreSQL 10
* ``local-tar`` backups are now split to roughly 2 gigabyte chunks of
  plain-text files.  New basebackups now benefit from some parallelization,
  as up to 5 chunks may be uploaded in parallel while new backup chunks are
  being created.
* ``pghoard_restore`` can now download and extract chunked basebackups using
  all CPU cores of the machine assuming that the parallel download mechanism
  can deliver it chunks fast enough
* Alternative ``pghoard_postgres_command`` implementation in Go for faster
  startup times.  The time that the Python implementation takes to start
  limits WAL restore throughput quite a bit
* Remove support for ``pg_xlog_directory`` config variable, we now require
  ``pg_data_directory`` to be set in the config for the backup site.  This
  is done in order to stop guessing and possibly getting the answer wrong
  for the real location of the data.
* Miscellaneous bug fixes

pghoard 1.5.0 (2017-06-07)
==========================

* New tool, ``pghoard_archive_cleanup`` to clean up any orphan WAL segments
  from the object store
* Basebackup ``end-time`` and ``end-wal-segment`` are now stored in metadata and
  used for PITR when ``local-tar`` basebackups are used
* Azure object storage updated to work with the latest python-azure-storage
  module and promoted to production ready status
* Google object storage driver now retries operations on failure to work
  around transient connection reset and backend failure issues
* Google object storage compatibility with Google oauth2client version 1.5+
* S3 driver now supports server side encryption
* Swift driver now supports the region_name configuration option
* Support for ``local-tar`` backups of a replica on PostgreSQL 9.6 and newer
  using the new non-exclusive backup mechanism.  Previous PostgreSQL
  versions require the pgespresso extension to take backups of replicas
  using the ``local-tar`` method, the pg_basebackup utilizing (default)
  methods have always supported backups of replicas
* When ``local-tar`` basebackups are used in exclusive mode (PG <= 9.5 without
  pgespresso) any conflicting exclusive basebackup is automatically cancelled
  to allow PGHoard to take its own basebackup.  This is sometimes required if
  a previous PGHoard process has been killed while it was taking a backup
* Support pghoard_restore --recovery-target-action with 9.3 and 9.4
* Creating a ``maintenance_mode_file`` as documented in README now stops
  automatic basebackups from being created
* Changed the default directory for ``maintenance_mode_file`` and
  ``json_state_file_path`` from ``/tmp`` to ``/var/lib/pghoard``
* Autotune ``compression.thread_count`` and ``transfer_agent.thread_count`` defaults
  to be ``max(cpu_count, 5)`` instead of a default of ``5`` as before
* Support for Python 3.6
* Miscellaneous bug fixes

pghoard 1.4.0 (2016-07-22)
==========================

* Add ability to get state file from webserver using ``GET /status``
* Support for Telegraf and DataDog statsd
* Basebackup restoration now shows download progress
* New ``basebackup_mode`` option: ``local-tar`` to collect files directly from
  ``$PGDATA`` instead of running pg_basebackup.  The ``local-tar`` mode allows
  backing up user tablespaces
* New site-specific configuration option ``pg_data_directory`` is required
  to use the new ``local-tar`` mode, but it's also recommended in other
  configurations as the ``pg_xlog_directory`` option is now deprecated and
  will be removed in a future release
* New site-specific configuration option ``pg_bin_directory`` replaces the
  previous global ``pg_basebackup_path`` and ``pg_receivexlog_path``

pghoard/config.py
"""
pghoard - configuration validation

Copyright (c) 2016 Ohmu Ltd
See LICENSE for details
"""
import json
import multiprocessing
import os
import subprocess
from typing import Optional

from rohmu import get_class_for_transfer
from rohmu.errors import InvalidConfigurationError
from rohmu.snappyfile import snappy

from pghoard.common import (extract_pg_command_version_string, pg_major_version, pg_version_string_to_number)
from pghoard.postgres_command import PGHOARD_HOST, PGHOARD_PORT

SUPPORTED_VERSIONS = ["14", "13", "12", "11", "10", "9.6", "9.5", "9.4", "9.3"]


def get_cpu_count():
    return multiprocessing.cpu_count()


def get_command_version(command: str, can_fail=True) -> Optional[str]:
    """
    Run the given command identified by it's full path with the --version
    option.
    """
    if os.path.exists(command) and os.access(command, os.X_OK):
        try:
            version_output = subprocess.check_output([command, "--version"])
            version_string = version_output.decode("ascii").strip()
            version_string = extract_pg_command_version_string(version_string)
            return version_string
        except subprocess.CalledProcessError:
            if not can_fail:
                raise
    elif not can_fail:
        raise FileNotFoundError(f"{command} doesn't exist")
    return None


def find_pg_binary(wanted_program, versions=None, pg_bin_directory=None, check_commands=True):
    """
    Find pg binary tries to find the wanted_program in one of the wanted
    versions using the following locations:
     - if the pg_bin_directory is provided, try that first
     - if we can't find it in the pg_bin_directory, look for "well known
       locations", which are where RPM / DEB packages install multiple versions
       of postgresql
    - if we still can't find it in those locations, fall back to examining the
      path.
    """
    if wanted_program in ("pg_receivexlog", "pg_receivewal"):
        programs = ["pg_receivexlog", "pg_receivewal"]
    else:
        programs = [wanted_program]
    pathformats = ["/usr/pgsql-{ver}/bin/{prog}", "/usr/lib/postgresql/{ver}/bin/{prog}"]
    if pg_bin_directory is not None:
        pathformats = [pg_bin_directory + "/{prog}"]
    versions = versions or SUPPORTED_VERSIONS
    for ver in versions:
        for pathfmt in pathformats:
            for program in programs:
                command = pathfmt.format(ver=ver, prog=program)
                if os.path.exists(command):
                    if check_commands is False:
                        return command, ver
                    command_version = get_command_version(command)
                    if pg_major_version(command_version) == ver:
                        return command, command_version
    # We couldn't find a supported version in the "well known locations",
    # let's search in the path.
    for path in os.environ["PATH"].split(os.pathsep):
        for program in programs:
            command = os.path.join(path, program)
            if not os.path.exists(command):
                continue
            version_string = get_command_version(command)
            if version_string and pg_major_version(version_string) in versions:
                return command, version_string
    raise RuntimeError("Cannot find program(s): {}".format(", ".join(programs)))


def set_and_check_config_defaults(config, *, check_commands=True, check_pgdata=True):
    # TODO: consider implementing a real configuration schema at some point
    # misc global defaults
    config.setdefault("backup_location", None)
    config.setdefault("http_address", PGHOARD_HOST)
    config.setdefault("http_port", PGHOARD_PORT)
    config.setdefault("alert_file_dir", config.get("backup_location") or os.getcwd())
    config.setdefault("json_state_file_path", "/var/lib/pghoard/pghoard_state.json")
    config.setdefault("maintenance_mode_file", "/var/lib/pghoard/maintenance_mode_file")
    config.setdefault("log_level", "INFO")
    config.setdefault("path_prefix", "")  # deprecated, used in the default path for sites
    config.setdefault("tar_executable", "pghoard_gnutaremu")
    config.setdefault("upload_retries_warning_limit", 3)
    config.setdefault("hash_algorithm", "sha1")

    # default to cpu_count + 1 compression threads
    config.setdefault("compression", {}).setdefault(
        "thread_count",
        get_cpu_count() + 1,
    )
    # default to cpu_count + 3 transfer threads (max 20)
    config.setdefault("transfer", {}).setdefault(
        "thread_count",
        min(get_cpu_count() + 3, 20),
    )
    # With 20 processes the restoration is almost always IO bound so using more CPU
    # cores isn't typically useful and just adds general overhead.
    # Only create CPU count + 1 processes as hosts with small number of CPUs often
    # have little memory too and each restore process can use fair amount of memory.
    config.setdefault("restore_process_count", min(get_cpu_count() + 1, 20))
    # default to prefetching transfer.thread_count objects
    config.setdefault("restore_prefetch", config["transfer"]["thread_count"])
    # if compression algorithm is not explicitly set prefer snappy if it's available
    if snappy is not None:
        config["compression"].setdefault("algorithm", "snappy")
    else:
        config["compression"].setdefault("algorithm", "lzma")
    config["compression"].setdefault("level", 0)

    # defaults for sites
    config.setdefault("backup_sites", {})
    for site_name, site_config in config["backup_sites"].items():
        site_config.setdefault("active", True)
        site_config.setdefault("active_backup_mode", "pg_receivexlog")

        site_config.setdefault("basebackup_chunk_size", 1024 * 1024 * 1024 * 2)
        site_config.setdefault("basebackup_chunks_in_progress", 5)
        site_config.setdefault("basebackup_compression_threads", 0)
        site_config.setdefault("basebackup_count", 2)
        site_config.setdefault("basebackup_count_min", 2)
        site_config.setdefault("basebackup_delta_mode_max_retries", 10)
        # Group delta files which are smaller than this value into chunks
        site_config.setdefault("basebackup_delta_mode_min_delta_file_size", 1024 * 1024 * 10)
        site_config.setdefault("basebackup_delta_mode_chunk_size", 1024 * 1024 * 500)
        site_config.setdefault("basebackup_interval_hours", 24)
        # NOTE: stream_compression removed from documentation after 1.6.0 release
        site_config.setdefault("basebackup_mode", "pipe" if site_config.get("stream_compression") else "basic")
        site_config.setdefault("basebackup_threads", 1)
        site_config.setdefault("encryption_key_id", None)
        site_config.setdefault("object_storage", None)
        pg_receivexlog_config = site_config.setdefault("pg_receivexlog", {})
        pg_receivexlog_config.setdefault("disk_space_check_interval", 10.0)
        pg_receivexlog_config.setdefault("resume_multiplier", 1.5)
        site_config.setdefault("prefix", os.path.join(config["path_prefix"], site_name))

        # NOTE: pg_data_directory doesn't have a default value
        data_dir = site_config.get("pg_data_directory")
        if not data_dir and check_pgdata:
            raise InvalidConfigurationError("Site {!r}: pg_data_directory must be set".format(site_name))

        if check_pgdata:
            version_file = os.path.join(data_dir, "PG_VERSION") if data_dir else None
            with open(version_file, "r") as fp:
                site_config["pg_data_directory_version"] = fp.read().strip()

        obj_store = site_config["object_storage"] or {}
        if not obj_store:
            pass
        elif "storage_type" not in obj_store:
            raise InvalidConfigurationError("Site {!r}: storage_type not defined for object_storage".format(site_name))
        elif obj_store["storage_type"] == "local" and obj_store.get("directory") == config.get("backup_location"):
            raise InvalidConfigurationError(
                "Site {!r}: invalid 'local' target directory {!r}, must be different from 'backup_location'".format(
                    site_name, config.get("backup_location")
                )
            )
        else:
            try:
                get_class_for_transfer(obj_store)
            except ImportError as ex:
                raise InvalidConfigurationError(
                    "Site {0!r} object_storage: {1.__class__.__name__!s}: {1!s}".format(site_name, ex)
                )
        fill_config_command_paths(config, site_name, check_commands)
    return config


def fill_config_command_paths(config, site_name, check_commands):
    # Set command paths and check their versions per site.  We use a configured value if one was provided
    # (either at top level or per site), if it wasn't provided but we have a valid pg_data_directory with
    # PG_VERSION in it we'll look for commands for that version from the expected paths for Debian and
    # RHEL/Fedora PGDG packages or otherwise fall back to iterating over the available versions.
    # Instead of setting paths explicitly for both commands, it's also possible to just set the
    # pg_bin_directory to point to the version-specific bin directory.
    site_config = config["backup_sites"][site_name]
    bin_dir = site_config.get("pg_bin_directory")
    for command in ["pg_basebackup", "pg_receivexlog"]:
        # NOTE: pg_basebackup_path and pg_receivexlog_path removed from documentation after 1.6.0 release
        command_key = "{}_path".format(command)
        command_path = site_config.get(command_key) or config.get(command_key)
        version_int = None
        if not command_path:
            pg_versions_to_check = None
            needs_check = check_commands and site_config["active"]
            if "pg_data_directory_version" in site_config:
                pg_versions_to_check = [site_config["pg_data_directory_version"]]
            try:
                command_path, version_string = find_pg_binary(command, pg_versions_to_check, bin_dir, needs_check)
                version_int = pg_version_string_to_number(version_string)
            except RuntimeError as _:
                # Only raise an error if we are expected to validate
                # the commands. Otherwise let it fail later
                if needs_check:
                    raise InvalidConfigurationError(
                        "Site {!r} command {!r} not found from path {}".format(site_name, command, command_path)
                    )
        elif check_commands and site_config["active"]:
            version_string = get_command_version(command_path, can_fail=False)
            version_int = pg_version_string_to_number(version_string)
        site_config[command_key] = command_path
        site_config[command + "_version"] = version_int


def read_json_config_file(filename, *, check_commands=True, add_defaults=True, check_pgdata=True):
    try:
        with open(filename, "r") as fp:
            config = json.load(fp)
    except FileNotFoundError:
        raise InvalidConfigurationError("Configuration file {!r} does not exist".format(filename))
    except ValueError as ex:
        raise InvalidConfigurationError("Configuration file {!r} does not contain valid JSON: {}".format(filename, str(ex)))
    except OSError as ex:
        raise InvalidConfigurationError(
            "Configuration file {!r} can't be opened: {}".format(filename, ex.__class__.__name__)
        )

    if not add_defaults:
        return config

    return set_and_check_config_defaults(config, check_commands=check_commands, check_pgdata=check_pgdata)


def get_site_from_config(config, site):
    if not config.get("backup_sites"):
        raise InvalidConfigurationError("No backup sites defined in configuration")
    site_count = len(config["backup_sites"])
    if site is None:
        if site_count > 1:
            raise InvalidConfigurationError(
                "Backup site not set and configuration file defines {} sites: {}".format(
                    site_count, sorted(config["backup_sites"])
                )
            )
        site = list(config["backup_sites"])[0]
    elif site not in config["backup_sites"]:
        n_sites = "{} other site{}".format(site_count, "s" if site_count > 1 else "")
        raise InvalidConfigurationError(
            "Site {!r} not defined in configuration file.  {} are defined: {}".format(
                site, n_sites, sorted(config["backup_sites"])
            )
        )

    return site


def key_lookup_for_site(config, site):
    def key_lookup(key_id):
        return config["backup_sites"][site]["encryption_keys"][key_id]["private"]

    return key_lookup
